### PR TITLE
feat: add layer zero examples

### LINKED ISSUE #3339: LZ <-> Hedera examples
### Problem

The PR check requires the linked issue to be from the current repository. The entire description of the tasks is here https://github.com/hashgraph/hedera-smart-contracts/issues/1030.

### Solution

Add LZ - Hedera examples.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by natanasow**
**Description**:

LZ is deployed on Hedera and we'd like to dig into see what's possible when considering HTS and it's representation fo ERC20 and ERC 721s.

Explore the LZ offering, and gain familiarity as a team and POC so cross-chain bridging of HTS tokens

**Self-improvement docs**:
- [LZ docs](https://docs.layerzero.network/v2)
- [OApp](https://docs.layerzero.network/v2/developers/evm/oapp/overview)
- [OFT](https://docs.layerzero.network/v2/developers/evm/oft/quickstart)
- [ONFT](https://docs.layerzero.network/v2/developers/evm/onft/quickstart)
- [LZ EVM Contracts](https://github.com/LayerZero-Labs/LayerZero-v2/tree/main/packages/layerzero-v2/evm)
- [LZ Notion Link](https://limechain.notion.site/Layer-Zero-Hedera-15a10baa8dcd8020a77edd60f41816c7)

### Tasks
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1060
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1077
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1078
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1079
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1080
- [x] https://github.com/hashgraph/hedera-smart-contracts/issues/1061

**Related issue(s)**:

Fixes #3339

**Notes for reviewer**:

**Checklist**

- [ ] Documented (Code comments, README, etc.)
- [ ] Tested (unit, integration, etc.)

### CODE REVIEW

**Review by quiet-node — COMMENTED** [review]
Without context it's quite hard to follow the PR. Could you provide a design document or any documentation that outlines the purpose of the project? The document could also define terms such as OFT and OApp. I believe the document you shared with the team last Friday was very effective. Would you consider updating and including it in this project?

File: tools/layer-zero-example/hardhat.config.js [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+};
+
+const getEndpointAddress = (network) => {
+  let ENDPOINT_V2;
+
+  if (network === 'hedera_testnet') {
+    ENDPOINT_V2 = '0xbD672D1562Dd32C23B563C989d8140122483631d';
+  } else if (network === 'bsc_testnet') {
+    ENDPOINT_V2 = '0x6EDCE65403992e310A62460808c4b910D972f10f';
+  }
```
Comment (quiet-node) [inline tools/layer-zero-example/hardhat.config.js:61]: What are these hard-coded contracts? How to get these? 
Maybe a comment explain these values would be valuable for devs working on or revisiting this in future.

File: tools/layer-zero-example/contracts/ExampleOFTHTS.sol [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+        uint32 _dstEid
+    ) internal virtual override returns (uint256 amountSentLD, uint256 amountReceivedLD) {
+        (amountSentLD, amountReceivedLD) = _debitView(_amountLD, _minAmountLD, _dstEid);
+
+        address spender = _msgSender();
+        if (_from != spender) {
+            int256 response = HederaTokenService.approve(htsTokenAddress, spender, amountSentLD);
+            require(response == HederaResponseCodes.SUCCESS, "HTS: Approve failed");
+        }
+
```
Comment (Ferparishuertas) [inline tools/layer-zero-example/contracts/ExampleOFTHTS.sol:72]: Isnt the transferFrom missing to be able to make a burn?

       int256 transferResponse = HederaTokenService.transferToken(htsTokenAddress,from,address(this) int64(uint64(_amountLD)));
        require(transferResponse == HederaResponseCodes.SUCCESS, "HTS: Transfer failed");

File: tools/layer-zero-example/contracts/ExampleOFTHTS.sol [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    function _debit(
+        address _from,
+        uint256 _amountLD,
+        uint256 _minAmountLD,
+        uint32 _dstEid
+    ) internal virtual override returns (uint256 amountSentLD, uint256 amountReceivedLD) {
+        (amountSentLD, amountReceivedLD) = _debitView(_amountLD, _minAmountLD, _dstEid);
+
+        address spender = _msgSender();
+        if (_from != spender) {
```
Comment (Ferparishuertas) [inline tools/layer-zero-example/contracts/ExampleOFTHTS.sol:68]: Goof extra. Could be with external approve too

File: tools/layer-zero-example/hardhat.config.js (hunk 1 above)
Comment (natanasow) [reply tools/layer-zero-example/hardhat.config.js:61]: They are described here https://docs.layerzero.network/v2/developers/evm/technical-reference/deployed-contracts. Yeah, the PR review requires some LZ knowledge :confused:. I will add the notion link to the description here and write down some comments to the codebase. Thanks!

File: tools/layer-zero-example/contracts/ExampleOFTHTS.sol (hunk 3 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/ExampleOFTHTS.sol]: Completely removed the approval from this call. It's user's/dapp's responsibility to ensure the token's approval beforehand like it's done in ERC20 swaps via 2 transactions.

File: tools/layer-zero-example/hardhat.config.js (hunk 1 above)
Comment (quiet-node) [reply tools/layer-zero-example/hardhat.config.js:61]: Yes, please add resources and comments for additional context. Otherwise, revisiting this or anyone not familiar LZ knowledge look into it might cause confusion. Thanks much!

Comment (natanasow) [reply tools/layer-zero-example/hardhat.config.js:61]: Added.

File: tools/layer-zero-example/README.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+npx hardhat test --grep "HTSAdapterTests @hedera @send" --network hedera_testnet
+npx hardhat test --grep "HTSAdapterTests @bsc @send" --network bsc_testnet
+
+wait a couple minutes, the LZ progress can be tracked on https://testnet.layerzeroscan.com/tx/<tx_hash>
+
+npx hardhat test --grep "HTSAdapterTests @hedera @test" --network hedera_testnet
+npx hardhat test --grep "HTSAdapterTests @bsc @test" --network bsc_testnet
+
+### WHBAR flow
```
Comment (Nana-EC) [inline tools/layer-zero-example/README.md:153]: This shouldn't be in the layer zero dir in it's final form.
Move this to `tools/whbar`

File: tools/layer-zero-example/contracts/HTSConnector.sol [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+     * @dev Burns tokens from the sender's specified balance.
+     * @param _from The address to debit the tokens from.
+     * @param _amountLD The amount of tokens to send in local decimals.
+     * @param _minAmountLD The minimum amount to send in local decimals.
+     * @param _dstEid The destination chain ID.
+     * @return amountSentLD The amount sent in local decimals.
+     * @return amountReceivedLD The amount received in local decimals on the remote.
+     */
+    function _debit(
+        address _from,
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/HTSConnector.sol:94]: Q: If `_from` is not msg.sender then it presumes an approval exists for the contract right?
If not do we need a check?

File: tools/layer-zero-example/contracts/WHBAR.sol [hunk 6]
```diff
@@ -0,0 +1,72 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.22;
+
+contract WHBAR {
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/WHBAR.sol:4]: nit: let's add a comment that this is based on WETH9 with updated solidity version and functions

File: tools/layer-zero-example/contracts/hts/HederaResponseCodes.sol [hunk 7]
```diff
@@ -0,0 +1,330 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.4.9 <0.9.0;
+
+library HederaResponseCodes {
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/hts/HederaResponseCodes.sol:4]: nit: might not be necessary to dupe the whole file. 
Why not just pull in the response code you need and add a comment ref to the HederaResponseCodes in the smart contracts repo.
Once we get the npm package this will be easier

File: tools/layer-zero-example/contracts/hts/IHederaTokenService.sol [hunk 8]
```diff
@@ -0,0 +1,799 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.4.9 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+interface IHederaTokenService {
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/hts/IHederaTokenService.sol:5]: Lots of methods and potentially a few structs that might not be needed to define here.
Again you can link in comment to `IHederaTokenService` in the smart contracts repo.
Once we get the npm package going ti'll be easier.

File: tools/layer-zero-example/contracts/hts/KeyHelper.sol [hunk 9]
```diff
@@ -0,0 +1,182 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity >=0.5.0 <0.9.0;
+pragma experimental ABIEncoderV2;
+
+import "./HederaTokenService.sol";
+
+abstract contract KeyHelper {
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/hts/KeyHelper.sol:7]: Same as other, reduce to only what's needed for this example and link in comment to the larger one

File: tools/layer-zero-example/hardhat.config.js [hunk 10]
```diff
@@ -0,0 +1,224 @@
...
+      accounts: [process.env.BSC_PK]
+    },
+  }
+};
+
+const getEndpointAddress = (network) => {
+  let ENDPOINT_V2;
+
+  if (network === 'hedera_testnet') {
+    ENDPOINT_V2 = '0xbD672D1562Dd32C23B563C989d8140122483631d';
```
Comment (Nana-EC) [inline tools/layer-zero-example/hardhat.config.js:58]: Q: is this the official LZ testnet contract or one you pushed for this example.
Either was note this in case the user hits the testnet after a reset.
In which case they'll either need to push one for their tests or point to the LZ one.
You should note this in the README and make sure to link to the endpoint code

File: tools/layer-zero-example/README.md [hunk 11]
```diff
@@ -0,0 +1,175 @@
+# Layer Zero examples
+
+### OApp
```
Comment (Nana-EC) [inline tools/layer-zero-example/README.md:3]: You should add a bit of a description to each section or ahead of the commands so it's clear what you're doing in the setup and what the test runs cover

**Review by Nana-EC — COMMENTED** [review]
Nice progress, some suggestions

File: tools/layer-zero-example/contracts/HTSConnector.sol (hunk 5 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/HTSConnector.sol:94]: There was a check here and approval that had been executed afterward but after some feedback from Fernando, I decided to completely remove the approval from this call. It's user's/dapp's responsibility to ensure the token's approval beforehand like it's done in ERC20 swaps via 2 transactions so we don't need any extra actions here.

File: tools/layer-zero-example/contracts/WHBAR.sol (hunk 6 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/WHBAR.sol:4]: Done.

File: tools/layer-zero-example/contracts/hts/HederaResponseCodes.sol (hunk 7 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/hts/HederaResponseCodes.sol:4]: Done.

File: tools/layer-zero-example/contracts/hts/IHederaTokenService.sol (hunk 8 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/hts/IHederaTokenService.sol:5]: Done.

File: tools/layer-zero-example/contracts/hts/KeyHelper.sol (hunk 9 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/hts/KeyHelper.sol:7]: Done.

File: tools/layer-zero-example/README.md (hunk 11 above)
Comment (natanasow) [reply tools/layer-zero-example/README.md:3]: Added.

File: tools/layer-zero-example/README.md (hunk 4 above)
Comment (natanasow) [reply tools/layer-zero-example/README.md:237]: Agreed. I put it there because the WHBAR example is on another branch and if I create a duplication of it in that PR, there will be future conflicts to resolve. Will move it as you suggested after I finalize the README.

File: tools/layer-zero-example/contracts/HTSConnector.sol [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _lzEndpoint,
+        address _delegate
+    ) payable OFTCore(8, _lzEndpoint, _delegate) {
+        IHederaTokenService.TokenKey[] memory keys = new IHederaTokenService.TokenKey[](2);
+        keys[0] = getSingleKey(
+            KeyType.ADMIN,
+            KeyType.PAUSE,
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/HTSConnector.sol:34]: I don't think we need the pause key. Let's keep this part simple with just admin and supply.

File: tools/layer-zero-example/contracts/HTSConnector.sol [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+     */
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _lzEndpoint,
+        address _delegate
+    ) payable OFTCore(8, _lzEndpoint, _delegate) {
+        IHederaTokenService.TokenKey[] memory keys = new IHederaTokenService.TokenKey[](2);
+        keys[0] = getSingleKey(
+            KeyType.ADMIN,
```
Comment (Nana-EC) [inline tools/layer-zero-example/contracts/HTSConnector.sol:33]: Wondering if we even need the admin key or only the supply. 
I forget if you have to specify admin in this case or if we can make the token immutable.
Even if we set it shouldn't it be a `CONTRACT_ID` type?

File: tools/layer-zero-example/hardhat.config.js [hunk 14]
```diff
@@ -0,0 +1,226 @@
...
+  return ENDPOINT_V2;
+}
+
+task('deploy-whbar', "Deploy WHBAR")
+  .setAction(async (taskArgs, hre) => {
+    const contractFactory = await ethers.getContractFactory('WHBAR');
+    const contract = await contractFactory.deploy();
+    await contract.deployTransaction.wait();
+
+    console.log(`(${hre.network.name}) WHBAR to: ` + contract.address);
```
Comment (Nana-EC) [inline tools/layer-zero-example/hardhat.config.js:74]: Can you log the tx hashes in these tasks.
Would be nice to follow the transactions in an E2E run

File: tools/layer-zero-example/test/htsAdapterTests.js [hunk 15]
```diff
@@ -0,0 +1,139 @@
...
+const { Options, addressToBytes32 } = require('@layerzerolabs/lz-v2-utilities');
+const { expect } = require('chai');
+
+const HEDERA_EID = 40285;
+const BSC_EID = 40102;
+const receiverAddress = '0xF51c7a9407217911d74e91642dbC58F18E51Deac';
+const amount = '100';
+
+describe('HTSAdapterTests', function() {
+  it('@hedera @fund-and-approve transfer to adapter', async () => {
```
Comment (Nana-EC) [inline tools/layer-zero-example/test/htsAdapterTests.js:32]: Similar to other comment can you log all the state changing transaction hashes for all the tests to make it easy to tell the E2E story on a given run

**Review by Nana-EC — COMMENTED** [review]
Nice work.
Some suggestions to make it easy to illustrate the E2E flow

File: tools/layer-zero-example/contracts/HTSConnector.sol [hunk 16]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    /**
+     * @dev Retrieves the address of the underlying ERC20 implementation.
+     * @return The address of the OFT token.
+     *
+     * @dev In the case of OFT, address(this) and erc20 are the same contract.
+     */
+    function token() public view returns (address) {
+        return address(this);
```
[the remark this answers is not in the document]
Comment (natanasow) [reply tools/layer-zero-example/contracts/HTSConnector.sol]: Fixed.

File: tools/layer-zero-example/contracts/HTSConnector.sol (hunk 12 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/HTSConnector.sol:34]: Removed.

File: tools/layer-zero-example/contracts/HTSConnector.sol (hunk 13 above)
Comment (natanasow) [reply tools/layer-zero-example/contracts/HTSConnector.sol:33]: Removed.

File: tools/layer-zero-example/test/htsAdapterTests.js (hunk 15 above)
Comment (natanasow) [reply tools/layer-zero-example/test/htsAdapterTests.js:32]: Done.

**Review by Nana-EC — APPROVED** [review]
LG, in general we've got the coverage we were looking for.
There's a few tweaks and additions we can make but i'd prefer this PR doesn't get anymore scope creep so let's get this in and address other improvements in follow up since these flows are giving the E2E examples we had hoped for

**Review by rbarkerSL — APPROVED** [review]

**Review by jeromy-cannon — APPROVED** [review]

File: tools/layer-zero-example/contracts/ExampleOApp.sol [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+            payable(msg.sender)
+        );
+    }
+
+    /**
+     * @dev Called when data is received from the protocol. It overrides the equivalent function in the parent contract.
+     * Protocol messages are defined as packets, comprised of the following parameters.
+     * @param _origin A struct containing information about where the packet came from.
+     * @param _guid A global unique identifier for tracking the packet.
+     * @param payload Encoded message.
```
Comment (quiet-node) [inline tools/layer-zero-example/contracts/ExampleOApp.sol:43]: Just a small nit: the doc doesn't have address and callData. This could be addressed in later PR not this one.

File: tools/layer-zero-example/test/htsConnectorTests.js [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+ */
+
+const hre = require('hardhat');
+const { ethers } = hre;
+const { Options, addressToBytes32 } = require('@layerzerolabs/lz-v2-utilities');
+const { expect } = require('chai');
+
+const HEDERA_EID = 40285;
+const BSC_EID = 40102;
+const receiverAddress = '0xF51c7a9407217911d74e91642dbC58F18E51Deac';
```
Comment (quiet-node) [inline tools/layer-zero-example/test/htsConnectorTests.js:28]: Same in next PR let's leave a link about the EIDs, and a comment where the receierAddress is from.

**Review by quiet-node — APPROVED** [review]
Great work! LGTM. Just some nits and questions which can be addressed in next PRs

File: tools/layer-zero-example/test/htsAdapterTests.js [hunk 19]
```diff
@@ -0,0 +1,143 @@
...
+ */
+
+const hre = require('hardhat');
+const { ethers } = hre;
+const { Options, addressToBytes32 } = require('@layerzerolabs/lz-v2-utilities');
+const { expect } = require('chai');
+
+const HEDERA_EID = 40285;
+const BSC_EID = 40102;
+const receiverAddress = '0xF51c7a9407217911d74e91642dbC58F18E51Deac';
```
[the remark this answers is not in the document]
Comment (Nana-EC) [reply tools/layer-zero-example/test/htsAdapterTests.js:28]: Closing to unblock merge, @natanasow to resolve in follow up

File: tools/layer-zero-example/contracts/ExampleOApp.sol (hunk 17 above)
Comment (Nana-EC) [reply tools/layer-zero-example/contracts/ExampleOApp.sol:43]: Closing to unblock merge, @natanasow to resolve in follow up

File: tools/layer-zero-example/test/htsConnectorTests.js (hunk 18 above)
Comment (Nana-EC) [reply tools/layer-zero-example/test/htsConnectorTests.js:28]: Closing to unblock merge, @natanasow to resolve in follow up
